rfd/0019-event-iteration-api.md

```diff
@@ -0,0 +1,110 @@
+---
+authors: Joel Wejdenstal (jwejdenstal@goteleport.com)
+state: draft
+---
+
+# RFD 19 - Event Fetch API with Pagination
+
+## What 1
+
+Implement a new API based on the current API for fetching events but with full pagination support in form of a `StartKey: string` parameter that allows the client to search from a specific point in the event stream instead of starting at the first event of the start date.
+
+## Why 2
+
+This RFC was sparked from issue [#5435](https://github.com/gravitational/teleport/issues/5435).
+The primary motivation for this RFD is that clients are at the moment not able to fetch an event stream between two dates incrementally with multiple API calls.
+
+Such functionality requires the client to be able to specify a point in the event stream from where the server will start searching forward from. Currently no such parameter exists which means the server will always start returning events from the start of the stream.
+
+## Details 3
+
+The current API for event fetch and search is over (https://github.com/gravitational/teleport/blob/master/lib/auth/apiserver.go#L246) plain HTTPS and not GRPC and runs on the authentication server.
+
+I propose implementing a new GRPC API that replaces and deprecates the old HTTP API endpoints. The old API endpoints should be kept as is but documentation should mention that it is deprecated and advise clients to use the new API.
+
+### GRPC Messages 3.1
+
+Out-of-session event request.
+This replaces and deprecates the `GET /:version/events` HTTP endpoint.
+
+These should be defined in [api/client/proto/authservice.proto](https://github.com/gravitational/teleport/blob/master/api/client/proto/authservice.proto).
+
+```protobuf
+message GetEventsRequest {
+   // Namespace, if not set, defaults to 'default'
+   Namespace string
+   // Oldest date of returned events
+   StartDate Timestamp
+   // Newest date of returned events
+   EndDate Timestamp
+   // EventType is optional, if not set, returns all events
+   EventType string
```

Comment:
right now the query is only limited to event types which is not useful by itself. Can we also search on event `JSON text` so people can find anything within that field? I think you can do that in dynamo DB.

Some use-case:

1. Want to find everything that happened on X server.
2. Want to find everything that had Y label.
3. Want to find all user Z events. 

I am suggesting to add a `Query` parameter that will be evaluated against the JSON text. This is in addition to `EventType`

Reply:
I'd push this out of scope. In DynamoDB this would activate full table scan and it will be timing out. Instead  I think in V2 we should really build a proper backend, like splunk or ES and revisit the query parameter.

Reply:
Agreed. This is way too much work to get working on current and future backends without some serious changes to schema too I believe. There's a lot of issues to figure out and architect around that I feel are vastly out of scope for what this RFD intends to fix and to deliver in a reasonable timeframe.